### PR TITLE
Get document stringification working properly for mutated inline documents

### LINKED ISSUE #315: Support cloning Documents, mutating them, and stringifying the results
The tricky part is inline documents. 

 - [x] `node-clone` needs some changes to be able to clone Documents.
   - [x] Clone maps and sets. We can probably get by with just this one for the time being. https://github.com/pvorb/node-clone/pull/64
   - [x] Clone promises. https://github.com/pvorb/node-clone/pull/65
   - [x] Clone symbols. https://github.com/pvorb/node-clone/pull/66
   - [x] Released as 2.0.0
 - [ ] Add AST nodes back onto Feature interface
 - [ ] Move serialization up to Document from ParsedDocument. Basic handling of serializing mutated inline documents.
 - [ ] Add optional indentation arg to serialize method.

### PR DESCRIPTION
**Pull request by rictic**
Fixes #315 

Also adds astNode back on Feature and ScannedFeature. This unblocks polymer-upgrade.

### CODE REVIEW

File: src/css/css-document.ts [hunk 1]
```diff
@@ -88,11 +88,17 @@ export class ParsedCssDocument extends ParsedDocument<shady.Node, Visitor> {
...
+  stringify(options?: StringifyOptions) {
+    options = options || {};
+    const beautifulResults =
+        cssbeautify(
+            shadyStringifier.stringify(this.ast),
+            {indent: '  ', autosemicolon: true, openbrace: 'end-of-line'}) +
         '\n';
+
+    return beautifulResults.split('\n')
+        .map(line => line && '  '.repeat(options.indent || 0) + line)
```
Comment (justinfagnani) [inline src/css/css-document.ts]: can you store the indent string in a variable?

Comment (justinfagnani) [inline src/css/css-document.ts]: when is `line` null? Does split() do that?

File: src/css/css-document.ts [hunk 2]
```diff
@@ -88,11 +88,17 @@ export class ParsedCssDocument extends ParsedDocument<shady.Node, Visitor> {
...
-  stringify() {
-    return cssbeautify(
-               shadyStringifier.stringify(this.ast),
-               {indent: '  ', autosemicolon: true, openbrace: 'end-of-line'}) +
+  stringify(options?: StringifyOptions) {
+    options = options || {};
+    const beautifulResults =
+        cssbeautify(
+            shadyStringifier.stringify(this.ast),
+            {indent: '  ', autosemicolon: true, openbrace: 'end-of-line'}) +
```
Comment (justinfagnani) [inline src/css/css-document.ts]: I assume cssbeautify has no option to indent an initial amount?

File: src/css/css-document.ts [hunk 3]
```diff
@@ -88,11 +88,17 @@ export class ParsedCssDocument extends ParsedDocument<shady.Node, Visitor> {
...
         '\n';
```
Comment (justinfagnani) [inline src/css/css-document.ts]: might be a little cleaner to at the final newline after the join?

File: src/html/html-document.ts [hunk 4]
```diff
@@ -102,21 +103,42 @@ export class ParsedHtmlDocument extends ParsedDocument<ASTNode, HtmlVisitor> {
...
-        result =
-            result.replace(RegExp(`<${tagName}>([^]*)?</${tagName}>`), '$1');
-      }
+    for (const doc of mutableDocuments) {
+      // TODO(rictic): infer this from doc.astNode's indentation.
+      const expectedIndentation = 2;
+
+      dom5.setTextContent(
+          doc.astNode,
+          '\n' + doc.stringify({indent: expectedIndentation}) + '  '.repeat(1));
```
Comment (justinfagnani) [inline src/html/html-document.ts]: TODO: infer the newline too?

Comment (justinfagnani) [inline src/html/html-document.ts]: `'  '.repeat(1)`?

File: src/html/html-document.ts [hunk 5]
```diff
@@ -102,21 +103,42 @@ export class ParsedHtmlDocument extends ParsedDocument<ASTNode, HtmlVisitor> {
     };
   }
 
-  stringify() {
-    // TODO(issues/315): serialize inline documents correctly.
-    let result = parse5.serialize(this.ast);
+  stringify(options?: StringifyOptions) {
+    options = options || {};
+    // We want to make of copy of `this` and all of the inline documents such
```
Comment (justinfagnani) [inline src/html/html-document.ts]: can you add why you need to clone? something about inline documents?

File: src/html/html-document.ts [hunk 6]
```diff
@@ -102,21 +103,42 @@ export class ParsedHtmlDocument extends ParsedDocument<ASTNode, HtmlVisitor> {
...
+  }
+}
+
+function prettyPrint(ast: dom5.Node, contents: string) {
+  let result = parse5.serialize(ast);
+
+  // Strip out inferred boilerplate nodes that are injected.
+  const injectedTagNames = ['html', 'head', 'body'];
+  for (const tagName of injectedTagNames) {
+    if (!contents.includes(`<${tagName}`)) {
```
Comment (justinfagnani) [inline src/html/html-document.ts]: I can already imaging funky things with import docs that contain strings or comments with `<body>`. Valdrin was showing me a bug related to that the other day. Can you check the source range of the ast nodes first, then remove html/head/body when they don't have a source range, and serialize?

File: src/css/css-document.ts (hunk 2 above)
Comment (rictic) [inline src/css/css-document.ts]: It does not :/

File: src/css/css-document.ts (hunk 3 above)
Comment (rictic) [inline src/css/css-document.ts]: Done.

File: src/css/css-document.ts (hunk 1 above)
Comment (rictic) [inline src/css/css-document.ts]: Done

Comment (rictic) [inline src/css/css-document.ts]: It's when it's the empty string, we don't want to splat out a bunch of trailing whitespace.

File: src/html/html-document.ts (hunk 5 above)
Comment (rictic) [inline src/html/html-document.ts]: Yeah, this is the craziest trick I've pulled in a little while. Expanded this comment out a bunch.

File: src/html/html-document.ts (hunk 4 above)
Comment (rictic) [inline src/html/html-document.ts]: I guess the thinking is that we only need newlines if the total content of this line is >80cols or whatever?

That's way more clever than I think we're likely to ever make this method, IMO.

Comment (rictic) [inline src/html/html-document.ts]: Fixed `'  '.repeat(1)` to something more sensible.

It made sense at the time, I promise!

File: src/html/html-document.ts (hunk 6 above)
Comment (rictic) [inline src/html/html-document.ts]: Really good point. I forgot that parser-hallucinated nodes wouldn't have __location info!

Done.

**Pull request merged**